Makefile

```diff
@@ -0,0 +1,40 @@
+SRC_FILES=src/ tests/ experiments/ examples/ docs/conf.py setup.py
```

Comment:
Do we need this in addition to `ci/code_checks.sh` and `.circleci/config.yml`?

With some work could perhaps be a replacement for `ci/code_checks.sh`, though it's missing some features right now (e.g. doing darglint but only on the diff).

It might actually be best to move to https://pre-commit.com/, I think set up right we could use that both locally and on CI (though this should be done in a separate PR if we do that!) Nora made a pre-commit config that's similar to what we do in `imitation` here: https://github.com/HumanCompatibleAI/go_attack/blob/main/.pre-commit-config.yaml

Reply:
Having a makefile is incredibly useful to run only specific commands without having to retype all the paths and settings every time, e.g. `make format` to format the file, or `make formatcheck` to check just formatting (and not typing). If I know typing passed before, I might want to only check formatting if the change I made is only e.g. a comma. 

Also, the makefile already prints out the commands that are run, like the options we have on the code checks bash script.

I think that we should 
1. Have a Makefile that contains a pre-commit command that properly replaces `./ci/code-checks.sh`
2. Automatically run the pre-commit command using the pre-commit config that you suggested
3. Point circleCI to the makefile if possible

For now I think there's no harm in leaving the Makefile in as that's what I use all the time instead of ./ci/code-checks.sh tbh.

Reply:
Having the checks defined in three places is going to be very brittle every time we change the CI checks. By all means use the Makefile personally but I'd rather it not be merged until it can replace `ci/code-checks.sh`.

I'm not sure I understand your concrete suggestion to use a `Makefile` and `pre-commit`. Isn't the Makefile redundant ifw we use `pre-commit`? Note you can specify a single hook to run https://pre-commit.com/#pre-commit-run so you'd still have support for just doing formatting, or just type-checking, etc.

Reply:
Okay, I didn't really read the docs of this tool properly. I was under the impression that you could only run it automatically and not manually. Yes, I agree 3 places for the checks is a bad idea. I just like the modularity of the `Makefile` which a shell script does not immediately provide without a lot of code overhead.

Reply:
I think there's stuff in the Makefile that does not really make sense to include in the pre-commit hook, such as frequently-used development scripts, e.g. make clean removes cache and other folders that are not committed to the repo. In this example, this is not a useful pre-commit action as git will ignore them anyways, but it's quite useful to have a `make clean` action. 

Probably worth moving this discussion off to another issue though, I don't think it should be a particularly hard thing to discuss but it seems useful to document the decision on its own.

Reply:
#554